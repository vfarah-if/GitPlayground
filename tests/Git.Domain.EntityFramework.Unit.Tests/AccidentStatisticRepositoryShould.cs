using FluentAssertions;
using System;
using System.Data.Entity;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using ApprovalTests;
using ApprovalTests.Reporters;
using Git.Domain.Models.TFL;
using Xunit;

namespace Git.Domain.EntityFramework.Unit.Tests
{
    [UseReporter(typeof(DiffReporter))]
    public class AccidentStatisticRepositoryShould : IDisposable
    {
        private readonly IAccidentStatisticDbContext _accidentStatisticDbContext;
        private readonly AccidentStatisticRepository _subject;

        public AccidentStatisticRepositoryShould()
        {
            _accidentStatisticDbContext = new AccidentStatisticDbContext();
            _subject = new AccidentStatisticRepository(_accidentStatisticDbContext);
            int actualCount = _accidentStatisticDbContext.AccidentStatistics.Count();
            if (actualCount == 0)
            {
                // Hack to let the temp database generate in the background
                // TODO: Come up with a more efficient way of doing this
                do
                {
                    Thread.Sleep(10000);
                } while (!_accidentStatisticDbContext.AccidentStatistics.Any());
            }
        }

        [Fact]
        public async Task ContainDataForThisIntegrationTestToRunSuccessfully()
        {
            var actualCount = await _accidentStatisticDbContext.AccidentStatistics.CountAsync();

            actualCount.Should().BeGreaterThan(0, "If this test fails than there is no database generated yet");
        }

        [Fact]
        public async Task CountAccidentStatistics()
        {
            var actualCount = await _subject.Count();

            actualCount.Should().BeGreaterThan(0);
        }

        [Fact]
        public async Task CountForOneAccidentStatisticWhenFilteringForTheFirstAccident()
        {
            var actualCount = await _subject.Count( x=> x.AccidentStatisticId == 1);

            actualCount.Should().Be(1);
        }
        
        [Fact]        
        public void GetAListOfFilteredByFatalCyclingAccidentsAndSortedByIdAscending()
        {
            var actual = _subject.Get(filter => 
                    filter.Severity == Severity.Fatal &&
                    filter.Casualties.Any(casualty =>
                        casualty.Mode.Equals("PedalCycle") && 
                        casualty.Severity == Severity.Fatal)
                ,sort => sort.AccidentStatisticId, true).GetAwaiter().GetResult();

            Approvals.VerifyJson(actual.ToJson());
        }

        [Fact]
        public void GetAListOfFilteredByFatalCyclingAccidentsAndSortedByIdDescending()
        {
            var actual = _subject.Get(filter =>
                    filter.Severity == Severity.Fatal &&
                    filter.Casualties.Any(casualty =>
                        casualty.Mode.Equals("PedalCycle") &&
                        casualty.Severity == Severity.Fatal)
                , sort => sort.AccidentStatisticId, false)
                .GetAwaiter()
                .GetResult();

            Approvals.VerifyJson(actual.ToJson());
        }

        [Fact]
        public void GetAListOfFilteredByFatalCyclingAccidentsAndSortedByIdDescendingAndOnlyReturnPage1Of1()
        {
            var actual = _subject.Get(filter =>
                    filter.Severity == Severity.Fatal &&
                    filter.Casualties.Any(casualty =>
                        casualty.Mode.Equals("PedalCycle") &&
                        casualty.Severity == Severity.Fatal)
                , sort => sort.AccidentStatisticId, false, 1, 1)
                .GetAwaiter()
                .GetResult();

            Approvals.VerifyJson(actual.ToJson());
        }


        [Fact]
        public void GetAListOfFilteredByFatalCyclingAccidentsAndSortedByIdDescendingAndOnlyReturnPage2Of1()
        {
            var actual = _subject.Get(filter =>
                    filter.Severity == Severity.Fatal &&
                    filter.Casualties.Any(casualty =>
                        casualty.Mode.Equals("PedalCycle") &&
                        casualty.Severity == Severity.Fatal)
                , sort => sort.AccidentStatisticId, false, 2, 1)
                .GetAwaiter()
                .GetResult();

            Approvals.VerifyJson(actual.ToJson());
        }

        [Fact]
        public void GetAListOfFilteredByFatalCyclingAccidentsAndWhenThePageIsHigherThan20ReturnPage20()
        {
            const int AboveTwentyPage = 21;
            var actual = _subject.Get(filter =>
                        filter.Severity == Severity.Fatal &&
                        filter.Casualties.Any(casualty =>
                            casualty.Mode.Equals("PedalCycle") &&
                            casualty.Severity == Severity.Fatal)
                    , sort => sort.AccidentStatisticId, false, AboveTwentyPage, 1)
                .GetAwaiter()
                .GetResult();

            Approvals.VerifyJson(actual.ToJson());
        }

        [Theory]
        [InlineData(0)]
        [InlineData(-1)]
        public void GetAListOfFilteredByFatalCyclingAccidentsAndWhenThePageIsBelowZeroShouldReturnTheFirstPage(
            int invalidPage)
        {
            var actual = _subject.Get(filter =>
                        filter.Severity == Severity.Fatal &&
                        filter.Casualties.Any(casualty =>
                            casualty.Mode.Equals("PedalCycle") &&
                            casualty.Severity == Severity.Fatal)
                    , sort => sort.AccidentStatisticId, false, invalidPage, 1)
                .GetAwaiter()
                .GetResult();

            Approvals.VerifyJson(actual.ToJson());
        }

        public void Dispose()
        {
            _accidentStatisticDbContext.Dispose();
        }
    }
}
