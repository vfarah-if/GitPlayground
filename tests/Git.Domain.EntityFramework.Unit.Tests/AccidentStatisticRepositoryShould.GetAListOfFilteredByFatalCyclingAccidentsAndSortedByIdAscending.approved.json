{
    
  "data":  [
        
    {
            
      "casualties":  [
                
        {
                    
          "casualtyId":  800,
                    
          "age":  32,
                    
          "class":  "Driver",
                    
          "severity":  2,
                    
          "mode":  "PedalCycle",
                    
          "ageBand":  "Adult"
        
                }
      
            ],
            
      "vehicles":  [
                
        {
                    
          "vehicleId":  1213,
                    
          "vehicleType":  "BusOrCoach"
        
                },
                
        {
                    
          "vehicleId":  1214,
                    
          "vehicleType":  "PedalCycle"
        
                }
      
            ],
            
      "accidentStatisticId":  714,
            
      "id":  297972,
            
      "lat":  "51.516166",
            
      "lon":  "-0.069496",
            
      "location":  "White Church Lane junction with Whitechapel Road",
            
      "date":  "2017-02-06T08:55:00",
            
      "severity":  2,
            
      "borough":  "Tower Hamlets"
    
        },
        
    {
            
      "casualties":  [
                
        {
                    
          "casualtyId":  863,
                    
          "age":  36,
                    
          "class":  "Driver",
                    
          "severity":  2,
                    
          "mode":  "PedalCycle",
                    
          "ageBand":  "Adult"
        
                }
      
            ],
            
      "vehicles":  [
                
        {
                    
          "vehicleId":  1308,
                    
          "vehicleType":  "HeavyGoodsVehicle"
        
                },
                
        {
                    
          "vehicleId":  1309,
                    
          "vehicleType":  "PedalCycle"
        
                }
      
            ],
            
      "accidentStatisticId":  771,
            
      "id":  305372,
            
      "lat":  "51.485807",
            
      "lon":  "-0.149825",
            
      "location":  "Grosvenor Road junction with Chelsea Bridge Road",
            
      "date":  "2017-09-27T08:15:00",
            
      "severity":  2,
            
      "borough":  "Kensington and Chelsea"
    
        },
        
    {
            
      "casualties":  [
                
        {
                    
          "casualtyId":  2992,
                    
          "age":  51,
                    
          "class":  "Driver",
                    
          "severity":  2,
                    
          "mode":  "PedalCycle",
                    
          "ageBand":  "Adult"
        
                }
      
            ],
            
      "vehicles":  [
                
        {
                    
          "vehicleId":  4475,
                    
          "vehicleType":  "MediumGoodsVehicle"
        
                },
                
        {
                    
          "vehicleId":  4476,
                    
          "vehicleType":  "PedalCycle"
        
                }
      
            ],
            
      "accidentStatisticId":  2597,
            
      "id":  315519,
            
      "lat":  "51.510929",
            
      "lon":  "-0.319291",
            
      "location":  "Uxbridge Road junction with northfields Avenue",
            
      "date":  "2017-10-12T14:30:00",
            
      "severity":  2,
            
      "borough":  "Ealing"
    
        },
        
    {
            
      "casualties":  [
                
        {
                    
          "casualtyId":  5003,
                    
          "age":  78,
                    
          "class":  "Driver",
                    
          "severity":  2,
                    
          "mode":  "PedalCycle",
                    
          "ageBand":  "Adult"
        
                }
      
            ],
            
      "vehicles":  [
                
        {
                    
          "vehicleId":  7418,
                    
          "vehicleType":  "HeavyGoodsVehicle"
        
                },
                
        {
                    
          "vehicleId":  7419,
                    
          "vehicleType":  "PedalCycle"
        
                }
      
            ],
            
      "accidentStatisticId":  4292,
            
      "id":  315618,
            
      "lat":  "51.509402",
            
      "lon":  "-0.376845",
            
      "location":  "South Road junction with St Josephs Drive",
            
      "date":  "2017-11-09T08:56:00",
            
      "severity":  2,
            
      "borough":  "Ealing"
    
        },
        
    {
            
      "casualties":  [
                
        {
                    
          "casualtyId":  10991,
                    
          "age":  32,
                    
          "class":  "Driver",
                    
          "severity":  2,
                    
          "mode":  "PedalCycle",
                    
          "ageBand":  "Adult"
        
                }
      
            ],
            
      "vehicles":  [
                
        {
                    
          "vehicleId":  15030,
                    
          "vehicleType":  "HeavyGoodsVehicle"
        
                },
                
        {
                    
          "vehicleId":  15031,
                    
          "vehicleType":  "PedalCycle"
        
                }
      
            ],
            
      "accidentStatisticId":  9348,
            
      "id":  308310,
            
      "lat":  "51.502723",
            
      "lon":  "0.023766",
            
      "location":  "North Woolwich Road junction with Knights Road",
            
      "date":  "2017-02-09T13:25:00",
            
      "severity":  2,
            
      "borough":  "Newham"
    
        },
        
    {
            
      "casualties":  [
                
        {
                    
          "casualtyId":  13588,
                    
          "age":  36,
                    
          "class":  "Driver",
                    
          "severity":  2,
                    
          "mode":  "PedalCycle",
                    
          "ageBand":  "Adult"
        
                }
      
            ],
            
      "vehicles":  [
                
        {
                    
          "vehicleId":  18800,
                    
          "vehicleType":  "LightGoodsVehicle"
        
                },
                
        {
                    
          "vehicleId":  18801,
                    
          "vehicleType":  "PedalCycle"
        
                }
      
            ],
            
      "accidentStatisticId":  11491,
            
      "id":  616855,
            
      "lat":  "51.549848",
            
      "lon":  "-0.128044",
            
      "location":  "Camden Road 20 metres west of junction with Hilldrop Crescent",
            
      "date":  "2017-08-29T06:50:00",
            
      "severity":  2,
            
      "borough":  "Islington"
    
        },
        
    {
            
      "casualties":  [
                
        {
                    
          "casualtyId":  19421,
                    
          "age":  30,
                    
          "class":  "Driver",
                    
          "severity":  2,
                    
          "mode":  "PedalCycle",
                    
          "ageBand":  "Adult"
        
                }
      
            ],
            
      "vehicles":  [
                
        {
                    
          "vehicleId":  28456,
                    
          "vehicleType":  "Car"
        
                },
                
        {
                    
          "vehicleId":  28457,
                    
          "vehicleType":  "Car"
        
                },
                
        {
                    
          "vehicleId":  28458,
                    
          "vehicleType":  "PedalCycle"
        
                }
      
            ],
            
      "accidentStatisticId":  16465,
            
      "id":  319343,
            
      "lat":  "51.628073",
            
      "lon":  "-0.049087",
            
      "location":  "Bounces Road junction with Exeter Road N9",
            
      "date":  "2017-02-06T01:42:00",
            
      "severity":  2,
            
      "borough":  "Enfield"
    
        },
        
    {
            
      "casualties":  [
                
        {
                    
          "casualtyId":  19849,
                    
          "age":  36,
                    
          "class":  "Driver",
                    
          "severity":  2,
                    
          "mode":  "PedalCycle",
                    
          "ageBand":  "Adult"
        
                }
      
            ],
            
      "vehicles":  [
                
        {
                    
          "vehicleId":  29096,
                    
          "vehicleType":  "LightGoodsVehicle"
        
                },
                
        {
                    
          "vehicleId":  29097,
                    
          "vehicleType":  "PedalCycle"
        
                }
      
            ],
            
      "accidentStatisticId":  16809,
            
      "id":  296665,
            
      "lat":  "51.549848",
            
      "lon":  "-0.128044",
            
      "location":  "Camden Road 20 metres west of junction with Hilldrop Crescent",
            
      "date":  "2017-08-29T06:50:00",
            
      "severity":  2,
            
      "borough":  "Islington"
    
        },
        
    {
            
      "casualties":  [
                
        {
                    
          "casualtyId":  26605,
                    
          "age":  33,
                    
          "class":  "Driver",
                    
          "severity":  2,
                    
          "mode":  "PedalCycle",
                    
          "ageBand":  "Adult"
        
                }
      
            ],
            
      "vehicles":  [
                
        {
                    
          "vehicleId":  39012,
                    
          "vehicleType":  "PedalCycle"
        
                }
      
            ],
            
      "accidentStatisticId":  22527,
            
      "id":  617890,
            
      "lat":  "51.558971",
            
      "lon":  "-0.053228",
            
      "location":  "Lea Bridge Road 15 metres south of junction with Cornthwaite Road",
            
      "date":  "2017-10-23T14:30:00",
            
      "severity":  2,
            
      "borough":  "Hackney"
    
        },
        
    {
            
      "casualties":  [
                
        {
                    
          "casualtyId":  29361,
                    
          "age":  32,
                    
          "class":  "Driver",
                    
          "severity":  2,
                    
          "mode":  "PedalCycle",
                    
          "ageBand":  "Adult"
        
                }
      
            ],
            
      "vehicles":  [
                
        {
                    
          "vehicleId":  48327,
                    
          "vehicleType":  "BusOrCoach"
        
                },
                
        {
                    
          "vehicleId":  48328,
                    
          "vehicleType":  "PedalCycle"
        
                }
      
            ],
            
      "accidentStatisticId":  24834,
            
      "id":  618162,
            
      "lat":  "51.516166",
            
      "lon":  "-0.069496",
            
      "location":  "White Church Lane junction with Whitechapel Road",
            
      "date":  "2017-02-06T08:55:00",
            
      "severity":  2,
            
      "borough":  "Tower Hamlets"
    
        },
        
    {
            
      "casualties":  [
                
        {
                    
          "casualtyId":  31702,
                    
          "age":  33,
                    
          "class":  "Driver",
                    
          "severity":  2,
                    
          "mode":  "PedalCycle",
                    
          "ageBand":  "Adult"
        
                }
      
            ],
            
      "vehicles":  [
                
        {
                    
          "vehicleId":  41806,
                    
          "vehicleType":  "PedalCycle"
        
                }
      
            ],
            
      "accidentStatisticId":  26825,
            
      "id":  297700,
            
      "lat":  "51.558971",
            
      "lon":  "-0.053228",
            
      "location":  "Lea Bridge Road 15 metres south of junction with Cornthwaite Road",
            
      "date":  "2017-10-23T14:30:00",
            
      "severity":  2,
            
      "borough":  "Hackney"
    
        },
        
    {
            
      "casualties":  [
                
        {
                    
          "casualtyId":  43702,
                    
          "age":  30,
                    
          "class":  "Driver",
                    
          "severity":  2,
                    
          "mode":  "PedalCycle",
                    
          "ageBand":  "Adult"
        
                }
      
            ],
            
      "vehicles":  [
                
        {
                    
          "vehicleId":  64090,
                    
          "vehicleType":  "Car"
        
                },
                
        {
                    
          "vehicleId":  64091,
                    
          "vehicleType":  "Car"
        
                },
                
        {
                    
          "vehicleId":  64092,
                    
          "vehicleType":  "PedalCycle"
        
                }
      
            ],
            
      "accidentStatisticId":  36726,
            
      "id":  639533,
            
      "lat":  "51.628073",
            
      "lon":  "-0.049087",
            
      "location":  "Bounces Road junction with Exeter Road N9",
            
      "date":  "2017-02-06T01:42:00",
            
      "severity":  2,
            
      "borough":  "Enfield"
    
        },
        
    {
            
      "casualties":  [
                
        {
                    
          "casualtyId":  44392,
                    
          "age":  57,
                    
          "class":  "Driver",
                    
          "severity":  2,
                    
          "mode":  "PedalCycle",
                    
          "ageBand":  "Adult"
        
                }
      
            ],
            
      "vehicles":  [
                
        {
                    
          "vehicleId":  65099,
                    
          "vehicleType":  "HeavyGoodsVehicle"
        
                },
                
        {
                    
          "vehicleId":  65100,
                    
          "vehicleType":  "PedalCycle"
        
                }
      
            ],
            
      "accidentStatisticId":  37291,
            
      "id":  625302,
            
      "lat":  "51.484078",
            
      "lon":  "-0.184216",
            
      "location":  "Fulham Road junction with Redcliffe Gardens",
            
      "date":  "2017-05-22T11:43:00",
            
      "severity":  2,
            
      "borough":  "Kensington and Chelsea"
    
        },
        
    {
            
      "casualties":  [
                
        {
                    
          "casualtyId":  48614,
                    
          "age":  36,
                    
          "class":  "Driver",
                    
          "severity":  2,
                    
          "mode":  "PedalCycle",
                    
          "ageBand":  "Adult"
        
                }
      
            ],
            
      "vehicles":  [
                
        {
                    
          "vehicleId":  71149,
                    
          "vehicleType":  "HeavyGoodsVehicle"
        
                },
                
        {
                    
          "vehicleId":  71150,
                    
          "vehicleType":  "PedalCycle"
        
                }
      
            ],
            
      "accidentStatisticId":  40686,
            
      "id":  625562,
            
      "lat":  "51.485807",
            
      "lon":  "-0.149825",
            
      "location":  "Grosvenor Road junction with Chelsea Bridge Road",
            
      "date":  "2017-09-27T08:15:00",
            
      "severity":  2,
            
      "borough":  "Kensington and Chelsea"
    
        },
        
    {
            
      "casualties":  [
                
        {
                    
          "casualtyId":  49027,
                    
          "age":  57,
                    
          "class":  "Driver",
                    
          "severity":  2,
                    
          "mode":  "PedalCycle",
                    
          "ageBand":  "Adult"
        
                }
      
            ],
            
      "vehicles":  [
                
        {
                    
          "vehicleId":  71757,
                    
          "vehicleType":  "HeavyGoodsVehicle"
        
                },
                
        {
                    
          "vehicleId":  71758,
                    
          "vehicleType":  "PedalCycle"
        
                }
      
            ],
            
      "accidentStatisticId":  41032,
            
      "id":  305112,
            
      "lat":  "51.484078",
            
      "lon":  "-0.184216",
            
      "location":  "Fulham Road junction with Redcliffe Gardens",
            
      "date":  "2017-05-22T11:43:00",
            
      "severity":  2,
            
      "borough":  "Kensington and Chelsea"
    
        },
        
    {
            
      "casualties":  [
                
        {
                    
          "casualtyId":  54822,
                    
          "age":  80,
                    
          "class":  "Driver",
                    
          "severity":  2,
                    
          "mode":  "PedalCycle",
                    
          "ageBand":  "Adult"
        
                }
      
            ],
            
      "vehicles":  [
                
        {
                    
          "vehicleId":  80136,
                    
          "vehicleType":  "HeavyGoodsVehicle"
        
                },
                
        {
                    
          "vehicleId":  80137,
                    
          "vehicleType":  "PedalCycle"
        
                }
      
            ],
            
      "accidentStatisticId":  45702,
            
      "id":  307574,
            
      "lat":  "51.575622",
            
      "lon":  "0.176232",
            
      "location":  "London Road junction with Knightsbridge Gardens",
            
      "date":  "2017-12-19T15:00:00",
            
      "severity":  2,
            
      "borough":  "Havering"
    
        },
        
    {
            
      "casualties":  [
                
        {
                    
          "casualtyId":  55918,
                    
          "age":  80,
                    
          "class":  "Driver",
                    
          "severity":  2,
                    
          "mode":  "PedalCycle",
                    
          "ageBand":  "Adult"
        
                }
      
            ],
            
      "vehicles":  [
                
        {
                    
          "vehicleId":  81696,
                    
          "vehicleType":  "HeavyGoodsVehicle"
        
                },
                
        {
                    
          "vehicleId":  81697,
                    
          "vehicleType":  "PedalCycle"
        
                }
      
            ],
            
      "accidentStatisticId":  46572,
            
      "id":  627764,
            
      "lat":  "51.575622",
            
      "lon":  "0.176232",
            
      "location":  "London Road junction with Knightsbridge Gardens",
            
      "date":  "2017-12-19T15:00:00",
            
      "severity":  2,
            
      "borough":  "Havering"
    
        },
        
    {
            
      "casualties":  [
                
        {
                    
          "casualtyId":  58654,
                    
          "age":  51,
                    
          "class":  "Driver",
                    
          "severity":  2,
                    
          "mode":  "PedalCycle",
                    
          "ageBand":  "Adult"
        
                }
      
            ],
            
      "vehicles":  [
                
        {
                    
          "vehicleId":  85601,
                    
          "vehicleType":  "MediumGoodsVehicle"
        
                },
                
        {
                    
          "vehicleId":  85602,
                    
          "vehicleType":  "PedalCycle"
        
                }
      
            ],
            
      "accidentStatisticId":  48769,
            
      "id":  635709,
            
      "lat":  "51.510929",
            
      "lon":  "-0.319291",
            
      "location":  "Uxbridge Road junction with northfields Avenue",
            
      "date":  "2017-10-12T14:30:00",
            
      "severity":  2,
            
      "borough":  "Ealing"
    
        },
        
    {
            
      "casualties":  [
                
        {
                    
          "casualtyId":  59792,
                    
          "age":  78,
                    
          "class":  "Driver",
                    
          "severity":  2,
                    
          "mode":  "PedalCycle",
                    
          "ageBand":  "Adult"
        
                }
      
            ],
            
      "vehicles":  [
                
        {
                    
          "vehicleId":  87193,
                    
          "vehicleType":  "HeavyGoodsVehicle"
        
                },
                
        {
                    
          "vehicleId":  87194,
                    
          "vehicleType":  "PedalCycle"
        
                }
      
            ],
            
      "accidentStatisticId":  49660,
            
      "id":  635808,
            
      "lat":  "51.509402",
            
      "lon":  "-0.376845",
            
      "location":  "South Road junction with St Josephs Drive",
            
      "date":  "2017-11-09T08:56:00",
            
      "severity":  2,
            
      "borough":  "Ealing"
    
        },
        
    {
            
      "casualties":  [
                
        {
                    
          "casualtyId":  60448,
                    
          "age":  32,
                    
          "class":  "Driver",
                    
          "severity":  2,
                    
          "mode":  "PedalCycle",
                    
          "ageBand":  "Adult"
        
                }
      
            ],
            
      "vehicles":  [
                
        {
                    
          "vehicleId":  88117,
                    
          "vehicleType":  "HeavyGoodsVehicle"
        
                },
                
        {
                    
          "vehicleId":  88118,
                    
          "vehicleType":  "PedalCycle"
        
                }
      
            ],
            
      "accidentStatisticId":  50185,
            
      "id":  628500,
            
      "lat":  "51.502723",
            
      "lon":  "0.023766",
            
      "location":  "North Woolwich Road junction with Knights Road",
            
      "date":  "2017-02-09T13:25:00",
            
      "severity":  2,
            
      "borough":  "Newham"
    
        }
  
    ],
    
  "page":  1,
    
  "lastPage":  1,
    
  "pageSize":  20,
    
  "total":  20

}